src/ssh/azext_ssh/custom.py

```diff
@@ -0,0 +1,115 @@
+# --------------------------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for license information.
+# --------------------------------------------------------------------------------------------
+
+import functools
+import os
+import hashlib
+import json
+
+from knack import util
+
+from . import ip_utils
+from . import rsa_parser
+from . import ssh_utils
+
+
+def ssh_vm(cmd, resource_group_name=None, vm_name=None, ssh_ip=None, public_key_file=None, private_key_file=None):
+    _do_ssh_op(cmd, resource_group_name, vm_name, ssh_ip,
+               public_key_file, private_key_file, ssh_utils.start_ssh_connection)
+
+
+def ssh_config(cmd, config_path, resource_group_name=None, vm_name=None, ssh_ip=None,
+               public_key_file=None, private_key_file=None):
+    op_call = functools.partial(ssh_utils.write_ssh_config, config_path, resource_group_name, vm_name)
+    _do_ssh_op(cmd, resource_group_name, vm_name, ssh_ip, public_key_file, private_key_file, op_call)
+
+
+def _do_ssh_op(cmd, resource_group, vm_name, ssh_ip, public_key_file, private_key_file, op_call):
+    _assert_args(resource_group, vm_name, ssh_ip)
+    public_key_file, private_key_file = _check_public_private_files(public_key_file, private_key_file)
+    ssh_ip = ssh_ip or ip_utils.get_ssh_ip(cmd, resource_group, vm_name)
+
+    if not ssh_ip:
+        raise util.CLIError(f"VM '{vm_name}' does not have a public IP address to SSH to")
+
+    scopes = ["https://pas.windows.net/CheckMyAccess/Linux/user_impersonation"]
+    data = _prepare_jwk_data(public_key_file)
+    from azure.cli.core._profile import Profile
+    profile = Profile(cli_ctx=cmd.cli_ctx)
+    username, certificate = profile.get_msal_token(scopes, data)
+
+    cert_file = _write_cert_file(public_key_file, username, certificate)
+    op_call(ssh_ip, username, cert_file, private_key_file)
+
+
+def _prepare_jwk_data(public_key_file):
```

Comment:
These codes were previously in azure-cli-core. I move it here because we think it is ssh certificate specific. In azure-cli-core, we expose a function `get_msal_token`  in _profile.Profile to make it simple and clean. @rlrossiter @danybeam, could you please help review it ?

Reply:
I don't like the use of the `_profile` module from core outside of core because modules that start with _ usually denote that it's a private module that shouldn't be used outside of the repo it lives in. That's why I put this in azure-cli-core to start

Reply:
Yes, I agree with you about the private module part. For this `_profile` case, as there are lots of examples in extensions to use `_profile` directly, so I think we can just follow it. From cli-core perspective, we have an internal discussion, and we think we'd better not include SSH specific logic in it.